### ISSUE TITLE
Compute from another Compute?

### ISSUE THREAD

**Issue by awei01** [cerebral/cerebral#288]
Is it possible to make Computed functions more composeable? 

Like if I have `computeFoo = Computed(something)` and `computeBasedOnFoo`. Is there a way to leverage the existing functionality of `computeFoo`? 

Or, is there a better coding pattern to handle this situation?

**Comment by christianalfoni** [comment]
@awei01 You can use a computed inside a computed:

```js
Computed({
  someOtherComputed: someOtherComputed()
}, props => {

})
```

That make sense?

**Comment by awei01** [comment]
yup. thanks.

**Comment by Guria** [comment]
@christianalfoni is it covered on website?

**Comment by awei01** [comment]
Hi, I realized what my issue was. I might have hit upon an edge case. Not sure if this is a bug or if it's supposed to be handled by `Computed`. Of course there are workarounds to get by this issue. But, thought you guys might be interested or have some insight to add. Or, I might be doing something incorrectly on my end. Please refer to this repo to illustrate: https://github.com/awei01/cerebral-computed-issue

Basically, if a `Computed` has an object as its first argument, it correctly updates whenever the underlying state changes. 

But, when a `Computed` is passed a function as the first argument, it seems to only run once and ignore underlying state changes.

**Comment by christianalfoni** [comment]
@Guria Do not think so, let me review and make update, thanks!

**Comment by christianalfoni** [comment]
@awei01 I believe you are right. The cache key we use does not take into account deps as a function. I will make a test and ensure this works, thanks for reporting! :)

**Comment by awei01** [comment]
Also, while we're at it, I had another question

Is it possible to use the result of a computed function to determine the result of another computed function. Something like:

```
const computeFoo = Computed({ 
   foo: 'path.to.foo'
}, ({ foo }) => {
  return something
})

const computeBasedOnFooAndParam = Computed((param) => {
    const foo = computeFoo() // this line obviously doesn't return what i'd expect
    const result = foo[param]
    return { result }
}, ({ result }) => {
  return something
})   


```

**Comment by christianalfoni** [comment]
@awei01 I just wrote a test for the computed issue you mentioned and it runs as it should, hm... not quite sure what differs in your scenario. Investigating further.

In your second case you have to:

```js
const computeFoo = Computed({ 
   foo: 'path.to.foo'
}, ({ foo }) => {
  return something
})

const computeBasedOnFooAndParam = Computed(({
  foo: computeFoo()
}, ({ foo, param }) => {
  return foo[param]
})  

computeBasedOnFooAndParam({param: 'woop'})
```

Any props you pass into the computed is also available in the callback :)

**Comment by christianalfoni** [comment]
Okay, I found the issue @awei01 . You have to pass arguments to computed as an object. It works just like props for a component, it merges in with the existing props. I will throw an error there if it is not an object :)

**Comment by awei01** [comment]
@christianalfoni Ok. It does work when I pass in an object. Thanks for spotting that. However, I didn't get a thrown Error. Maybe you just made that code change. 

Regarding the other question. I think my example was stupid and didn't really communicate my question. Let me try again:

```
const computeRoute = Compute({
   route: 'path.to.foo' 
}, (props) => {
  return props
})

const computeItemBasedOnKey = Compute((props) => {
   return {
      item: 'path.to.all.items.' + props.key
   }
}, (props) => {
   return props
})

const computeItemFromRoute = Compute(() => {
   const route = computeRoute()
   return computeItemBaseOnKey({ key: route.params.currentKey })
}, (props) => {
   return props
})

```

Also, maybe it's just me, but I also seem to be passing a simple `(props) => { return props }` as my second param for `Computed` a lot. Would it make sense to default `cb` is not passed, it can default to (props) => { return props }?

**Comment by christianalfoni** [comment]
Yeah, the error will throw in next release :)

Well, the concept of a computed is to depend on more than one path and use the values of those to produce a new "component friendly" value. There is not reason to do:

```js
const computeRoute = Compute({
   route: 'path.to.foo' 
}, (props) => {
  return props
})
```

Or this:

```js
const computeItemBasedOnKey = Compute((props) => {
   return {
      item: 'path.to.all.items.' + props.key
   }
}, (props) => {
   return props
})
```

The only thing you need here is:

```js
const computeItemFromRoute = Compute({
   currentKey: 'path.to.route.params.currentKey',
   items: 'path.to.items'
}, ({currentKey, items}) => {
   return items[currentKey]
})
```

From a component perspective you might rather want to do:

connect({
  currentKey: 'route.params.currentKey'
},
  function ItemsPage (props) {
    return (
      <div>
        {props.currentKey ? <Item itemKey={props.currentKey} /> : null}
      </div>
    )
  }
)

connect(props => {
  item: `items.${props.itemKey}`
},
  function Items (props) {
    return (
      <div>
       Woop woop: {props.item.title}
      </div>
    )
  }
)

**Comment by christianalfoni** [comment]
Btw, thanks for an extremely excellent bug report back there 👍 :D

**Comment by awei01** [comment]
Sure no prob on bug report. 

As for the computed of computed issue. I was trying to leverage existing computed functionality and not try to recompose keys as it has been done in a different compute. No worries. Keep up the great work.
